src/java_bytecode/java_string_library_preprocess.cpp

```diff
@@ -961,41 +962,24 @@ codet java_string_library_preprocesst::make_equals_function_code(
       constant_exprt("java::java.lang.String", string_typet()));
   }();
 
-  // THEN case
-  code_blockt instance_case;
-  {
+  code.then_case() = [&] {
+    code_blockt instance_case;
     // Check content equality
-    const exprt::operandst args =
-      process_equals_function_operands(ops, loc, symbol_table, instance_case);
     const symbolt string_equals_sym = get_fresh_aux_symbol(
-      java_boolean_type(),
-      "string_equals",
-      "str_eq",
-      loc,
-      ID_java,
-      symbol_table);
+      return_type, "string_equals", "str_eq", loc, ID_java, symbol_table);
     const symbol_exprt string_equals = string_equals_sym.symbol_expr();
     instance_case.add(code_declt(string_equals), loc);
-    instance_case.add(
-      code_assignt(
-        string_equals,
-        make_function_application(
-          ID_cprover_string_equal_func,
-          args,
-          type.return_type(),
-          symbol_table)),
-      loc);
+    const exprt::operandst args =
+      process_equals_function_operands(ops, loc, symbol_table, instance_case);
+    const auto fun_app = make_function_application(
+      ID_cprover_string_equal_func, args, return_type, symbol_table);
+    instance_case.add(code_assignt(string_equals, fun_app), loc);
     instance_case.add(code_returnt(string_equals), loc);
-  }
+    return instance_case;
+  }();
 
-  // ELSE case
-  const code_returnt non_instance_case(
-    from_integer(false, java_boolean_type()));
+  code.else_case() = code_returnt(from_integer(false, return_type));
```

Comment:
`code_returnt(false_exprt())`;

Reply:
`false_exprt` constructor does not take a type and has the wrong one for us (bool_typet instead of java_bool_typet)

Reply:
Ok, I wasn't aware that Java requires a different Boolean type.